src/index.ts

```diff
@@ -687,17 +807,22 @@ export function create (rawOptions: CreateOptions = {}): Register {
 
       // Set the file contents into cache manually.
       const updateMemoryCache = (contents: string, fileName: string) => {
-        const sourceFile = builderProgram.getSourceFile(fileName)
-
-        fileContents.set(fileName, contents)
+        const previousContents = fileContents.get(fileName)
+        const contentsChanged = previousContents !== contents
+        if (contentsChanged) {
+          fileContents.set(fileName, contents)
+        }
 
         // Add to `rootFiles` when discovered by compiler for the first time.
-        if (sourceFile === undefined) {
+        let addedToRootFileNames = false
+        if (!rootFileNames.has(fileName) && !isFileKnownToBeInternal(fileName)) {
+          markBucketOfFilenameInternal(fileName)
           rootFileNames.add(fileName)
+          addedToRootFileNames = true
         }
 
         // Update program when file changes.
-        if (sourceFile === undefined || sourceFile.text !== contents) {
+        if (addedToRootFileNames || contentsChanged) {
           builderProgram = ts.createEmitAndSemanticDiagnosticsBuilderProgram(
             Array.from(rootFileNames),
```

Comment:
Just being a bit smarter about when a file must be added to `rootFileNames`.  Even if the compiler has already seen a file, it's possible it was marked external.